ENH: add getPVAliases and grep_more_ioc scripts

aberges-SLAC opened the pull request:
## Description
Added a script for reconstructing PV aliases from their child IOC's, `st.cmd` files, and `.archive` files for use in ArchiverAppliance scripts. Does a pythonic `grep_ioc` for the regex pattern supplied, then searches the child IOC for the parent's release pointer. After confirmation, reconstructs all the record ←→ aliases using `.../db/*.archive` and `st.cmd` from the child IOC.

Also added a script called `grep_more_ioc` which is an a personal utility inspired by `grep_ioc` . Most of the legwork of this script is reused in `getPVAlias` so I decided to bundle it for posterity. 

## Motivation and Context
Motors, namely SmarActs and Elliptecs, tend to move around to different areas often and/or have dynamic configurations for channels based on the area's application. This results in PV aliases becoming stale references to channels that no longer house that stage. To rectify this, you'll need to get the currently built association of EPICS records ←→ aliases for the PVs. 

Additionally, I added some QOL in `grep_more_ioc` that helps day-to-day controls work in CLI without having to open the `iocmanager` gui. 

## How Has This Been Tested?
I've run this script & wrapper in an ipython session and using the supplied bash wrapper. Please see the screenshots below!

## Where Has This Been Documented?
All documentation is supplied inline on the scripts with accompanying usage/help doctext.

## Screenshots (if appropriate):
### getPVAliases
Initial confirmation of ioc caught by a `grep_ioc` like search:
![image](https://github.com/pcdshub/engineering_tools/assets/149725219/beeda0a1-93f1-4ac9-b2d8-7fe3233a4793)

All axes/motors generated by the child IOC's `st.cmd` file:
![image](https://github.com/pcdshub/engineering_tools/assets/149725219/b6a01df0-a3b0-440e-b7af-74fbb203a123)

Preview of what will be saved to the `.txt` file dumps:
![image](https://github.com/pcdshub/engineering_tools/assets/149725219/2a5d9365-019e-4dcb-9b09-dcbe7ed33397)

Assumes current directory as the base for new folder generation & dump, but accepts different directories:
![image](https://github.com/pcdshub/engineering_tools/assets/149725219/3b752c25-5963-4e22-948e-86639183df81)

Makes the folder if it does not exist, then iterates to the next IOC:
![image](https://github.com/pcdshub/engineering_tools/assets/149725219/f2607cb0-ebea-4ea4-ae7e-78acdb8b42b9)

Individual axes can be skipped, with lazy skipping through the rest of the axes:
![image](https://github.com/pcdshub/engineering_tools/assets/149725219/313ea4e2-d0b0-44b5-8d77-c6690cd7aa96)

--> Otherwise if you don't accept _all_ axes dumped, (or lazy accept/skip the rest), you will be asked to confirm each dump:
![image](https://github.com/pcdshub/engineering_tools/assets/149725219/8ffc0550-e753-462c-aa62-06afe06e45d4)

Example of output:
![image](https://github.com/pcdshub/engineering_tools/assets/149725219/fd0c5f14-47f2-45b8-9b38-4c80e1bfc4e6)

---------------------------------------------------------------------------------------------------------------------------------
### grep_more_ioc

This script uses `pandas` to print DataFrames, which is quite slower than `prettytable` but doesn't suffer the wrapping/width issues that the latter does. (see below)
![image](https://github.com/pcdshub/engineering_tools/assets/149725219/0595e245-f346-4fd2-9a89-687f8b6537df)


There are 2 major functions: `print` and `search`. In my examples below I'll use the alias `gmi` for `grep_more_ioc`

#### print
Options:
- `skip_comments`
![image](https://github.com/pcdshub/engineering_tools/assets/149725219/9e9b5a21-6e89-4912-86a5-84f59c77c879)
Functions like `cat`, except you skip commented code. Useful for quickly grabbing IOC information when building `happi` entries.

- `release`
![image](https://github.com/pcdshub/engineering_tools/assets/149725219/9909a546-f4f2-4998-a751-0c2fd5aa6f33)
Searches the child IOCs for the release pointer, then adds an abbreviated path to the parent's current release.

- `print_dirs`
![image](https://github.com/pcdshub/engineering_tools/assets/149725219/85e2e0e4-7b75-4a77-ba2e-812a93eda3e3)
Dumps directories to terminal for quicker copy/pasting/`pushd`

#### search
Options:
- `quiet`
![image](https://github.com/pcdshub/engineering_tools/assets/149725219/9ca14ae3-486d-40dd-8b51-0060dbddee0d)
![image](https://github.com/pcdshub/engineering_tools/assets/149725219/1057d062-bb35-466c-bade-258cada92fab)
Suppress warnings for paths that don't exist when doing batch searches (as disabled IOCs are often previously deleted...)

- `only`
![image](https://github.com/pcdshub/engineering_tools/assets/149725219/e1739d26-604c-496f-a99c-0ba481d46f10)
Only print the search results, don't print the DataFrame.

tangkong commented on scripts/getPVAliases.sh:
```diff
@@ -0,0 +1,45 @@
+    exit 0
+fi
+
+# source into pcds_conda if not currently active
```
Were you running into issues sourcing `pcds_conda` after it's been already activated?  I've never had an issue doing so

aberges-SLAC replied:
It really came down to saving 2-3 seconds per execution in a different script that I based a lot of my ioc manipulation on. That script called `pandas` which is a notoriously long import so I was trying to shave time off as much as I could.

tangkong commented on scripts/getPVAliases.sh:
```diff
@@ -0,0 +1,45 @@
+EOF
+}
+
+# catch improper number of args or asking for help
```
I believe you can just let python's argparse handle bad args and the help request, if you pass all the arguments through to your python script.

aberges-SLAC replied:
`source pcds_conda` is greedy and eats `-h` :(

So I tossed this in there as a hacky work-around

aberges-SLAC replied:
Fixed with suggested changes

tangkong reviewed (commented):
Nice work, seems like this should be very helpful!  I've left a couple nitpicky notes, I'll take a bit to look at this from a higher level after this.  

* I like to avoid regex when possible.  It's powerful but often difficult to maintain and parse (by humans).  There's a few places here where I don't think you need it.  
* Thanks for the comments / docstrings!  Love to see it.

aberges-SLAC commented on scripts/getPVAliases.py:
```diff
@@ -0,0 +1,508 @@
+#%% Functions
+###############################################################################
+
+def simple_prompt(prompt: str, default: str = 'N'):
```
This is _extremely_ nice, and really which I looked for it before I reinvented the wheel, lol!

aberges-SLAC commented on scripts/getPVAliases.py:
```diff
@@ -0,0 +1,508 @@
+
+def find_ioc(hutch: str=None, patt: str=None):
+    """
+    A pythonic grep_ioc for gathering IOC details from the cfg file
```
I very much wanted to avoid `iocmanager`, but was stubborn about the `ioc.cfg` outputs which were just so very nearly `json`.  Doing it this way let me use regex liberally, although it should absolutely have restraint haha.

ZLLentz replied:
On the flip side, this code becomes much easier to maintain if we replicate the native parsing. I could imagine how all of the replace/strip/regex done here may run into some unknown edge case later down the line.

I personally would advocate for changing iocmanager's format to a real json structure but that feels out of scope here.

aberges-SLAC replied:
I'm a huge fan of that as well, but I agree it's outside of the scope for this script sadly.

I know regex is a spice and it's easy to over season, but I quite like the option of regex searching the innards of the `ioc.cfg` dicts as it can make it much easier for catching bulk info in the MODs, e.g. `lm2k2.*qmini`. Maybe this is also my laziness speaking and it's not a good reason to keep it in, but wouldn't the `execfile` route require you to know/provide the full ioc name explicitly? 

This is probably safer, tbh, since you can't accidentally grab info you don't want. But on the other hand, naming conventions for MCS2 controllers seldom include MODs/TILE info that is relevant (also, sometimes TILEs are stretch across 2 IOCs):
![image](https://github.com/pcdshub/engineering_tools/assets/149725219/3d25b1c6-2e82-4dbe-bb83-c7a34a350be2)

So for fixing all the HRM aliases you'd probably want to catch both these IOCs.

What's the smallest hammer I can use here for the same endgoal?

tangkong replied:
I think regex is probably a good way to tackle the `ioc.cfg` files, I think we're specifically proposing parsing the iocmanager config file with `execfile`.  Running `execfile` on the individual `ioc-*.cfg` doesn't itself make much sense.

ZLLentz commented on scripts/getPVAliases.sh:
```diff
@@ -0,0 +1,45 @@
+fi
+
+# execute python script
+python "$ENG_TOOLS_SCRIPTS/getPVAliases.py" $@
```
Feel free to also try something like this if sourcing pcds_conda is ever slow. It's likely that your script doesn't need the full environment sourced, and rather just needs the executable.

This also means you'd be able to do all the arg handling in the python script (if you removed source pcds_conda)
```suggestion
/cds/group/pcds/pyps/conda/py39/envs/pcds-5.8.2/bin/python "$ENG_TOOLS_SCRIPTS/getPVAliases.py" $@
```

aberges-SLAC replied:
I very much prefer this, and quite honestly, I'd rather do that and reduce the whole `getPVAliases.sh` to this, since all other arg handling is done by argparser and is now very easy to maintain.

ZLLentz commented on scripts/getPVAliases.py:
I'd recommend running all python scripts through `flake8` or `ruff` to check for python styling standards. If you're into having style automatically standardized for you there's also `black`.

I'm pointing this out here because I saw various minor flake8 violations as I was reading through the code.

tangkong replied:
Should we add a pre-commit config to `engineering_tools`? 🤔

aberges-SLAC replied:
Ah, quite a few offenders here in this script. Let me clean that up. I also left in the `#%%` from Spyder for cells that probably don't serve purpose for folks.

ZLLentz commented on scripts/getPVAliases.sh:
I'd recommend running all shell scripts through shellcheck to check for potential escaping issues etc.

ZLLentz commented on scripts/getPVAliases.py:
```diff
@@ -0,0 +1,508 @@
+        Returns the output as a list of dict after json formatting
+
+    """
+    hutch_list = ['xpp', 'xcs', 'cxi', 'mfx', 'mec', 'tmo', 'rix', 'xrt',
+		 'aux', 'det', 'fee', 'hpl', 'icl', 'las', 'lfe', 'kfe', 'tst',
+		 'thz', 'all']
```
These sorts of lists may get hard to manage over time.

I suggest either adding the missing `txi` or pulling this from the subdirectories present in the /cds/group/pcds/pyps/config directory.

aberges-SLAC replied:
Good catch! I think that's the safer option. I can just import that as a constant earlier in the script and refer to it and make it better to maintain.

ZLLentz commented on scripts/getPVAliases.py:
```diff
@@ -0,0 +1,508 @@
+    return list(set(_result))
+
+
+def search_file(file:str,output:list=None,patt:str=None,prefix:str='',
+                color_wrap:Fore=None):
```
I'm noticing that you've type hinted all the arguments but none of the return values, and none of the lists have their containing types, and a lot of these are type hinted incorrectly (type str set to None here, for example). Are you interested in filling in the missing type information or were you just going for a light pass to help write the script?

aberges-SLAC replied:
I can clean this up. I think I originally left some of these as default = `None` for catching missing args or forcing break points in earlier testing and forgot to revert. I should go around and fill in all the `list[str]` and `list[dict]` etc.

I keep `output : list[str] = None` for reconstructing the `grep_ioc` output that I then parse again, so maybe it's superfluous and I should be more clever about this.

ZLLentz commented on scripts/getPVAliases.sh:
```diff
@@ -0,0 +1,45 @@
+# execute python script
+python "$ENG_TOOLS_SCRIPTS/getPVAliases.py" $@
+
+exit
```
Is it customary to include a bare exit at the end of a script? What function does it serve? I'm worried this masks the return code from the python call.

aberges-SLAC replied:
This is probably just a bad habit from my end. No need to give it a bare `exit`, or any `exit` for that matter

ZLLentz reviewed (commented):
Script is good, put in my suggestions and nitpicks

tangkong reviewed (commented):
A quick once through, not paying too much attention to the logic (which I had assumed did not change much from my last review).  But after reading through I think I want to look at the details more closely

aberges-SLAC commented on scripts/getPVAliases.py:
```diff
@@ -0,0 +1,376 @@
+# %% Constants
+###############################################################################
+
+VALID_HUTCH = sorted([s for s
```
Yeah, I'm not entirely sure _how_ to get around that. I guess I could check in the walk if any valid `.cfg` files exist?

tangkong replied:
Checking the contents of the folder was my first instinct, but it's only marginally more stable.

aberges-SLAC replied:
Something to the effect of:
```
VALID_HUTCH = ['all']
for (path, dir, files) in os.walk('/cds/group/pcds/pyps/config'):
    if 'iocmanager.cfg' in files:
    VALID_HUTCH.append(path.rsplit(r'/', maxsplit=1)[-1])
sorted(VALID_HUTCH)
```

tangkong replied:
I was thinking something silly like
```python
import glob
sorted([d for d in glob.glob(path + '/*/') if glob.glob(d + 'iocmanager.cfg')])
```

aberges-SLAC replied:
Not silly at all! I like this much more. Very clean, much glob

aberges-SLAC commented on scripts/getPVAliases.py:
```diff
@@ -0,0 +1,376 @@
+                      default: str = os.getcwd(),) -> str:
+    """
+    Requests the user for a destination to save a file.
+    Tests if the resultant path exists and makes the directory if necessary.
```
It _used_ to, lol, my bad! I forgot to update the doctext on this function when I merged other versions together. Let me fix that in my next commit

aberges-SLAC commented on scripts/getPVAliases.py:
```diff
@@ -0,0 +1,376 @@
+def flatten_list(input_list: list[list]) -> list[str]:
+    """
+    Flatten a 2D lists and find its unique values.
+    Note: loses order due to set() call. sear
```
🙇

aberges-SLAC commented on scripts/getPVAliases.py:
```diff
@@ -0,0 +1,376 @@
+def build_table(input_data: list[dict], columns: list[str] = None,
+                **kwargs) -> PrettyTable:
+    """
+    Build a prettytable from a list of dicts/JSON.
```
I'm actually very partial to pandas for printing to terminal, but the import is a _slog_ and I wanted something a bit faster for this script. Now that I'm importing from `grep_more_ioc` anyway, kind of a moot point and I might as well just use pandas? The only slight annoyance is fixing column with to max alias size but that's trivial.

tangkong replied:
Not worth changing since it's here already, and you're doing some messing with the formatting.  I was just wondering.

aberges-SLAC commented on scripts/getPVAliases.py:
```diff
@@ -0,0 +1,376 @@
+# parser obj configuration
+parser = argparse.ArgumentParser(
+    prog='gatherPVAliases',
+    description="""gathers all record <-> alias associations from a child's
+     ioc.cfg, st.cmd, and parent ioc.cfg.""",
```
Interesting! I've never tried string concat, frankly, so it's nice to know that it works. Adding to next commit

aberges-SLAC commented on scripts/grep_more_ioc.py:
```diff
@@ -0,0 +1,511 @@
+"""A tool for enhancing the grep_ioc CLI tool for ECS-Delivery team
+enginer = aberges
```
That is mostly an artifact from my Spyder settings and honestly can be removed. The header docstring should probably be changed in both cases. Especially `grep_more_ioc` which used a different, almost arbitrary semantic versioning for my own notes/reference when it was a personal script.

tangkong replied:
Also the versions of these scripts will be captured in engineering tools' versions anyway

aberges-SLAC commented on scripts/grep_more_ioc.py:
```diff
@@ -0,0 +1,511 @@
+# %% Constants
+###############################################################################
+
+VALID_HUTCH = sorted([s for s
```
Sure! That works — I also could've just imported it from `grep_more_ioc` in the first place as it's used in the `find_ioc` function which depends on this constant.

aberges-SLAC commented on scripts/grep_more_ioc.py:
```diff
@@ -0,0 +1,511 @@
+                        help='Flag for excluding based'
+                        + ' on the "disabled" state.')
+    # subparsers
+    subparsers = (parser
+                  .add_subparsers(help='Follow-up commands after grep_ioc '
+                                  + 'executes'))
```
I'm one of those weirdos who prefers it, but it really only has utility when you're at 3+ methods (which, tbf, I should be avoiding in the first place). Added to next commit!

ZLLentz replied:
the `.something` on each new line is really nice when each step returns a thing that you can also call a method on, which happens more in some languages than in others. I don't see this pattern often in Python but it pays off in these situations:
```
my_thing
    .do_thing()
    .do_another_thing()
    .do_more_many_thing()
```

aberges-SLAC commented on scripts/grep_more_ioc.py:
```diff
@@ -0,0 +1,511 @@
+# %%% print
+# --------------------------------------------------------------------------- #
+    # print the dataframe
+    if hasattr(args, 'print'):
```
I think the sparse nature of argparse (:badumtiss:) including only the invoked parser & subparser made it weird for testing, and I got distracted by other things and didn't leave a note to clean up.

I should probably [do something with subparser.set_default](https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.set_defaults) to give me some namespace clues. Something like `subparser.set_default(command='print')` etc.

aberges-SLAC replied:
Turns out my original approach is one of the few stable ways to check for other possible `args`. When `argparse` parses the args, if any subparser is unused, none of its attributes get populated in the resultant `Namespace` object. 
![image](https://github.com/pcdshub/engineering_tools/assets/149725219/a2389126-e7bd-4f14-8812-63274a3a3ace)

I guess the other way of doing this would be generating a full `Namespace` object with all of the defaults set and to check the `bool` of any `sub-commands` because I already structured them as simple toggles at the top level. 
Something like:

```
class Args:
    self.patt = None
    self.hutch = None
    self.ignore_disabled = False
    ...
````

But then I'm hard coding all the defaults anyway that `argparse` was originally handling. 🤷

ZLLentz commented on scripts/getPVAliases.sh:
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/bash
+
+# execute python script
+/cds/group/pcds/pyps/conda/py39/envs/pcds-5.8.2/bin/python "$ENG_TOOLS_SCRIPTS/getPVAliases.py" "$@"
```
Two things here:

- Do you know where `$ENG_TOOLS_SCRIPTS` gets set? Mine doesn't even match the engineering tools on my path, so for me this script actually will run a different .py file than the one I intend!
- Maybe it'd be more portable to have this point to the same-directory .py file instead of using the custom environment variables

```suggestion
THIS_DIR="$(dirname "$(realpath "${BASH_SOURCE[0]}")")"
/cds/group/pcds/pyps/conda/py39/envs/pcds-5.8.2/bin/python "$THIS_DIR/getPVAliases.py" "$@"
```

ZLLentz commented:
I'm starting to feel like this has been stuck in review cycle for longer than it needs to be relative to how complete the code is, so I'm going to try to focus my reviews on bigger things related to deploying/sharing/using the scripts. I can't wait to see this in engineering tools proper.

ZLLentz commented on scripts/getPVAliases.py:
One additional big-picture thing: I recommend adding your new scripts to the readme.

aberges-SLAC commented:
I _just_ caught a bug that's worth fixing before merging as well. Happens when you `print -c` or `print -r` on the _very_ rare case where a `child_ioc.cfg` doesn't actually exist, but the `ioc` is absolutely valid and in `HUTCH/iocmanager.cfg` *cough PLCs cough*.

I'm really just commenting this here to remind myself to use `try/except` with `search_file` for handling these edge cases.
![image](https://github.com/pcdshub/engineering_tools/assets/149725219/1cdc7196-70f8-4633-b7c5-564cff53fb80)

aberges-SLAC commented:
Okay, I fixed the bug when searching IOCs that had `',\n history: [...'` causing the IOC to get skipped during the search. Added a sub-function for printing these `history` attributes within `grep_more_ioc` but I'll add that to the `README.md` in post, since the pre-commit hates the `README.md` syntax.

ZLLentz commented:
> the pre-commit hates the README.md syntax

What's the story here? Maybe we need to make some of the pre-commit entries ignore the readme?

aberges-SLAC commented:
> > the pre-commit hates the README.md syntax
> 
> What's the story here? Maybe we need to make some of the pre-commit entries ignore the readme?

It complained loudly at me for markdown syntax that I never edited, and then seemingly wanted to remove lines like `<tr/>` just to add them back. It was somewhere like 6-7 commits ago when I was struggling to get the pre-commit to accept me (turns out adding it to my local repo helps [:).

Just let it its thing and committed the pre-commit changes a few tries ago. Now that it is happy I'm hesitant to make any further modifications. Besides, its much cleaner on the web interface anyhow.

tangkong commented:
The README.md probably hadn't been pre-commit-checked, since our checks only run on modified files.

ZLLentz commented on README.md:
```diff
@@ -245,6 +245,28 @@ usage: grep_ioc KEYWORD [hutch]<br/>
+<tr>
+    <td>grep_more_ioc</td>
+    <td>
+usage: grep_more_ioc [-h] [-d] patt hutch {print,search} <br/>
```
The script is actually `grep_more_ioc.sh`, I think we should document as such here or rename the scripts to not have .sh suffixes.

silkenelson replied:
I'd vote to not have .sh as most scripts in engineering tools do not have extensions.

tangkong replied:
To hop on to the readme suggestions, could you add an example call for those of us who don't know what pattern/hutch to search for?   I think examples are helpful for people not so familiar with EPICS and our systems

ZLLentz replied:
I think if this script renaming/examples stuff get in then this PR has my approval (I don't want to hold it up any longer, but I'd like the names to be finalized + I'd like people to be ready to use it when it gets merged)

ZLLentz commented:
I ran the scripts myself and asked the questions that I had as a user, maybe they inform documentation/examples/etc.

ZLLentz commented on README.md:
```diff
@@ -245,6 +245,28 @@ usage: grep_ioc KEYWORD [hutch]<br/>
+    <td>
+usage: grep_more_ioc [-h] [-d] patt hutch {print,search} <br/>
+     positional arguments: <br/>
+     patt                            Regex str to search through iocmanager.cfg<br/>
+     hutch                           hutch to search<br/>
```
```
$ grep_more_ioc.sh --help
usage: grep_more_ioc [-h] [-d] patt hutch {print,search} ...

Transforms grep_ioc output to json object and prints in pandas.DataFrame

positional arguments:
  patt
  hutch
  {print,search}        Follow-up commands after grep_ioc executes
    print               Just a simple print of the dataframe to the terminal.
    search              For using regex-like searches in the child IOC.cfg captured by grep_ioc. Useful for quickly gathering instance information, IP addr, etc.
```

aberges-SLAC replied:
I can improve this in the `argparse` documentation so it is more transparent.

The _annoying_ thing about how `argparse` handles subparsers is that each `subparser` has its own help text, so something like:
`grep_more_ioc.sh . all print --help` triggers the help text for `print`
`grep_more_ioc.sh . all search --help` triggers the help text for `search`

Personally, I hate it, but I haven't found a good enough work around _yet_

tangkong replied:
One simple thing you could do is add some text in the top-level parser that hints the user to try those sub-parser help messages.

aberges-SLAC commented on scripts/getPVAliases.py:
```diff
@@ -0,0 +1,342 @@
+            print(build_table(alias_dicts, ['record', 'alias'], align='l'))
+            # optional skip for all resulting PV aliases
+            save_all = (simple_prompt(
+                'Do you want to save all resulting PV aliases? '
```
I should update this docstring to "Do you want to add all resulting PV <--> alias associations to the final output?"

The script now saves a single file with all the results appended together (less annoying to parse when you're playing with the result later). Since you're iterating through a base PV and its resulting PVs (each with aliases), instead of pestering the user to approve every single instance (e.g. each axis in an MCS2 ioc), the user can approve if the first association looks good.

ZLLentz replied:
I think if the script saves a file we should mention this in the docstring.

aberges-SLAC replied:
More informative?
![image](https://github.com/pcdshub/engineering_tools/assets/149725219/d6d6578a-2ea8-4615-9c45-08a19e9de668)

ZLLentz replied:
It's definitely better but I think my overall confusion was a lack of understanding that this script was supposed to save anything at all.

aberges-SLAC replied:
With some massaging on `argparse.RawTextHelpFormatter`:
![image](https://github.com/pcdshub/engineering_tools/assets/149725219/d70c8f44-3ae3-49a8-9a5a-2688a7d3a71a)

ZLLentz replied:
Looks good to me! Very helpful output!

aberges-SLAC commented on scripts/grep_more_ioc.py:
Yes, they are positional arguments. I contemplated making `grep_more_ioc.sh <patt> <hutch>` default to print but it gets hairy with the subparsers.

aberges-SLAC replied:
Also improved:
![image](https://github.com/pcdshub/engineering_tools/assets/149725219/29d4029f-408f-4e1c-bc04-e85ea8fa3ca1)

ZLLentz replied:
This one looks good too!

aberges-SLAC commented:
Okay, all help doc-text improved, committed, and pushed!

tangkong reviewed (dismissed):
This looks good to me.  I think future improvements probably deserve their own separate PRs.  This is some good work! 👍

ZLLentz reviewed (changes requested):
The final tiny thing before merge is the script names, the majority of the executables in engineering_tools don't have the .sh extension, so I suggest renaming these to remove it.

It's worth noting that you've already documented this without the .sh extension in the readme, so renaming things makes everything consistent.

Otherwise, if there's a good reason to keep the .sh, it should be documented as such in the readme.

aberges-SLAC commented:
> The final tiny thing before merge is the script names, the majority of the executables in engineering_tools don't have the .sh extension, so I suggest renaming these to remove it.
> 
> It's worth noting that you've already documented this without the .sh extension in the readme, so renaming things makes everything consistent.
> 
> Otherwise, if there's a good reason to keep the .sh, it should be documented as such in the readme.

Done and done! see new commits

ZLLentz reviewed (approved):
Let's get this in!

ZLLentz commented:
I'm going to merge this now. Post-merge if there are other concerns we can handle them in smaller, reviewable follow-up PRs. We can either tag/deploy this immediately or we can wait, let me know what you all prefer.

aberges-SLAC commented:
> I'm going to merge this now. Post-merge if there are other concerns we can handle them in smaller, reviewable follow-up PRs. We can either tag/deploy this immediately or we can wait, let me know what you all prefer.

I'm excited to get this out for folks to use & stress test. I'm sure there's more bugs hiding somewhere that's outside of my normal use cases. So I'm in favor of deployment sooner rather than later — the script is pretty stable for now.

Are there any confluences pages I should update as well? Or do we leave the documentation in the README.md?

ZLLentz commented:
I'll release another version right away then- tags are cheap. Do you want to announce the release of R3.8.0 (which, compared to last version, only has this PR) or would you prefer I do it?

I don't think there's any confluence page in particular that needs to be updated, unless you think a bigger page about usage etc. would be useful.
